File: src/nalgebra/cg.rs
//! Conjugate Gradient (CG) method.

use crate::{IterSolverError, IterSolverResult};
use nalgebra::{DMatrix, DVector};

/// Conjugate Gradient (CG) method for solving linear systems Ax = b.
///
/// The Conjugate Gradient method is an iterative algorithm for solving systems of linear
/// equations where the coefficient matrix A is symmetric and positive definite. It is
/// particularly effective for large sparse systems.
///
/// # Algorithm Overview
///
/// The CG method generates a sequence of approximations to the solution by minimizing
/// the quadratic form associated with the linear system. At each iteration, it computes
/// a search direction that is conjugate to all previous search directions with respect
/// to the matrix A.
///
/// # Convergence
///
/// For a symmetric positive definite matrix, the CG method is guaranteed to converge
/// to the exact solution in at most n iterations (where n is the size of the matrix),
/// though in practice it often converges much faster, especially for well-conditioned
/// systems.
///
/// # Examples
///
/// ```rust
/// use nalgebra::{DMatrix, DVector};
/// use iterative_solvers::CG;
///
/// // Create a simple 2x2 symmetric positive definite system
/// let mat = DMatrix::from_row_slice(2, 2, &[4.0, 1.0, 1.0, 3.0]);
/// let rhs = DVector::from_vec(vec![1.0, 2.0]);
/// let abstol = 1e-10;
/// let reltol = 1e-8;
///
/// let mut cg = CG::new(&mat, &rhs, abstol, reltol).unwrap();
/// let solution = cg.solve();
/// ```
#[derive(Debug, Clone)]
pub struct CG<'mat> {
    mat: &'mat DMatrix<f64>,
    solution: DVector<f64>,
    residual: f64,
    iteration: usize,
    r: DVector<f64>,
    c: DVector<f64>,
    u: DVector<f64>,
    tol: f64,
    prev_residual: f64,
}

impl<'mat> CG<'mat> {
    /// Create a new `CG` solver with the given matrix, right-hand side, and tolerance.
    ///
    /// # Arguments
    ///
    /// * `mat` - The coefficient matrix A.
    /// * `rhs` - The right-hand side vector b.
    /// * `abstol` - The absolute tolerance for the residual.
    /// * `reltol` - The relative tolerance for the residual.
    ///
    /// # Tolerance
    ///
    /// The tolerance is the stopping criterion for the CG method. It is the maximum of the absolute tolerance and the relative tolerance.
    ///
    /// The stopping criterion is `|r_k| ≤ max(reltol * |r_0|, abstol)`,
    /// where `r_k ≈ A x_k - b` is the residual at the `k`th iteration, and `r_0 ≈ A x_0 - b` is the initial residual.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use nalgebra::{DMatrix, DVector};
    /// use iterative_solvers::CG;
    ///
    /// // Create a simple 2x2 symmetric positive definite system
    /// let mat = DMatrix::from_row_slice(2, 2, &[4.0, 1.0, 1.0, 3.0]);
    /// let rhs = DVector::from_vec(vec![1.0, 2.0]);
    /// let abstol = 1e-10;
    /// let reltol = 1e-8;
    ///
    /// let mut cg = CG::new(&mat, &rhs, abstol, reltol).unwrap();
    /// let solution = cg.solve();
    /// ```
    pub fn new(
        mat: &'mat DMatrix<f64>,
        rhs: &'mat DVector<f64>,
        abstol: f64,
        reltol: f64,
    ) -> IterSolverResult<Self> {
        if !mat.is_square() {
            return Err(IterSolverError::DimensionError(format!(
                "The matrix is not square, whose shape is ({}, {})",
                mat.nrows(),
                mat.ncols()
            )));
        }
        if mat.nrows() != rhs.len() {
            return Err(IterSolverError::DimensionError(format!(
                "The matrix with order {}, and the rhs with length {}, do not match",
                mat.nrows(),
                rhs.len()
            )));
        }
        let n = mat.nrows();
        let x = DVector::zeros(n);
        let r = rhs.clone();
        let c = DVector::zeros(n);
        let u = DVector::zeros(n);
        let residual = r.norm();
        let prev_residual = residual;
        let iteration = 0;
        let tol = abstol.max(reltol * residual);
        Ok(Self {
            mat,
            solution: x,
            residual,
            iteration,
            r,
            c,
            u,
            tol,
            prev_residual,
        })
    }

    /// Create a new `CG` solver with the given matrix, right-hand side, tolerance, and initial guess.
    ///
    /// # Arguments
    ///
    /// * `mat` - The coefficient matrix A.
    /// * `rhs` - The right-hand side vector b.
    /// * `abstol` - The absolute tolerance for the residual.
    /// * `reltol` - The relative tolerance for the residual.
    /// * `initial_guess` - The initial guess for the solution.
    ///
    /// # Tolerance
    ///
    /// The tolerance is the stopping criterion for the CG method. It is the maximum of the absolute tolerance and the relative tolerance.
    ///
    /// The stopping criterion is `|r_k| ≤ max(reltol * |r_0|, abstol)`,
    /// where `r_k ≈ A x_k - b` is the residual at the `k`th iteration, and `r_0 ≈ A x_0 - b` is the initial residual.
    ///
    /// # Initial Guess
    ///
    /// The initial guess is the starting point for the CG method. It is used to compute the initial residual.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use nalgebra::{DMatrix, DVector};
    /// use iterative_solvers::CG;
    ///
    /// // Create a simple 2x2 symmetric positive definite system
    /// let mat = DMatrix::from_row_slice(2, 2, &[4.0, 1.0, 1.0, 3.0]);
    /// let rhs = DVector::from_vec(vec![1.0, 2.0]);
    /// let abstol = 1e-10;
    /// let reltol = 1e-8;
    /// let initial_guess = DVector::from_vec(vec![0.0, 0.0]);
    ///
    /// let mut cg = CG::new_with_initial_guess(&mat, &rhs, initial_guess, abstol, reltol).unwrap();
    /// let solution = cg.solve();
    /// ```
    pub fn new_with_initial_guess(
        mat: &'mat DMatrix<f64>,
        rhs: &'mat DVector<f64>,
        initial_guess: DVector<f64>,
        abstol: f64,
        reltol: f64,
    ) -> IterSolverResult<Self> {
        if !mat.is_square() {
            return Err(IterSolverError::DimensionError(format!(
                "The matrix is not square, whose shape is ({}, {})",
                mat.nrows(),
                mat.ncols()
            )));
        }
        if mat.nrows() != rhs.len() {
            return Err(IterSolverError::DimensionError(format!(
                "The matrix with order {}, and the rhs with length {}, do not match",
                mat.nrows(),
                rhs.len()
            )));
        }
        if initial_guess.len() != mat.nrows() {
            return Err(IterSolverError::DimensionError(format!(
                "The initial guess with length {}, and the matrix with order {}, do not match",
                initial_guess.len(),
                mat.nrows()
            )));
        }
        let n = mat.nrows();
        let x = initial_guess;
        let r = rhs - mat * &x;
        let c = DVector::zeros(n);
        let u = DVector::zeros(n);
        let residual = r.norm();
        let prev_residual = residual;
        let iteration = 0;
        let tol = abstol.max(reltol * residual);
        Ok(Self {
            mat,
            solution: x,
            residual,
            iteration,
            r,
            c,
            u,
            tol,
            prev_residual,
        })
    }

    /// Check if the solver has converged.
    #[inline]
    fn converged(&self) -> bool {
        self.residual <= self.tol
    }

    /// Check if the solver has reached the maximum number of iterations.
    #[inline]
    fn done(&self) -> bool {
        (self.iteration >= self.max_iter()) || self.converged()
    }

    /// Get the maximum number of iterations.
    #[inline]
    fn max_iter(&self) -> usize {
        self.solution.len()
    }

    /// Consume the solver and return the solved result.
    pub fn solve(mut self) -> Self {
        self.by_ref().count();
        self
    }

    /// Get the solution.
    pub fn solution(&self) -> &DVector<f64> {
        &self.solution
    }

    /// Get the residual.
    pub fn residual(&self) -> f64 {
        self.residual
    }

    /// Get the iteration.
    pub fn iteration(&self) -> usize {
        self.iteration
    }

    /// Get the matrix.
    pub fn mat(&self) -> &DMatrix<f64> {
        self.mat
    }

    /// Get the residual vector
    pub fn residual_vector(&self) -> &DVector<f64> {
        &self.r
    }

    /// Get the conjugate direction
    pub fn conjugate_direction(&self) -> &DVector<f64> {
        &self.u
    }
}

impl<'mat> Iterator for CG<'mat> {
    type Item = f64;

    fn next(&mut self) -> Option<Self::Item> {
        if self.done() {
            return None;
        }

        // u = r + beta * u
        let beta = self.residual.powi(2) / self.prev_residual.powi(2);
        self.u.axpy(1.0, &self.r, beta);

        // c = A * u
        self.mat.mul_to(&self.u, &mut self.c);

        // update solution and residual
        // x = x + alpha * u
        // r = r - alpha * c
        let alpha = self.residual.powi(2) / self.u.dot(&self.c);
        self.solution.axpy(alpha, &self.u, 1.0);
        self.r.axpy(-alpha, &self.c, 1.0);

        self.prev_residual = self.residual;
        self.residual = self.r.norm();

        self.iteration += 1;

        Some(self.residual)
    }
}

/// Solve a linear system Ax = b using the Conjugate Gradient method.
///
/// # Arguments
///
/// * `mat` - The coefficient matrix A.
/// * `rhs` - The right-hand side vector b.
/// * `abstol` - The absolute tolerance for the residual.
/// * `reltol` - The relative tolerance for the residual.
///
/// # Tolerance
///
/// The tolerance is the stopping criterion for the CG method. It is the maximum of the absolute tolerance and the relative tolerance.
///
/// The stopping criterion is `|r_k| ≤ max(reltol * |r_0|, abstol)`,
/// where `r_k ≈ A x_k - b` is the residual at the `k`th iteration, and `r_0 ≈ A x_0 - b` is the initial residual.
///
/// # Examples
///
/// ```rust
/// use nalgebra::{DMatrix, DVector};
/// use iterative_solvers::cg;
///
/// // Create a simple 2x2 symmetric positive definite system
/// let mat = DMatrix::from_row_slice(2, 2, &[4.0, 1.0, 1.0, 3.0]);
/// let rhs = DVector::from_vec(vec![1.0, 2.0]);
/// let abstol = 1e-10;
/// let reltol = 1e-8;
///
/// let solution = cg(&mat, &rhs, abstol, reltol).unwrap();
/// ```
pub fn cg<'mat>(
    mat: &'mat DMatrix<f64>,
    rhs: &'mat DVector<f64>,
    abstol: f64,
    reltol: f64,
) -> IterSolverResult<CG<'mat>> {
    let mut solver = CG::new(mat, rhs, abstol, reltol)?;
    solver.by_ref().count();
    Ok(solver)
}

/// Solve a linear system Ax = b using the Conjugate Gradient method.
///
/// # Arguments
///
/// * `mat` - The coefficient matrix A.
/// * `rhs` - The right-hand side vector b.
/// * `abstol` - The absolute tolerance for the residual.
/// * `reltol` - The relative tolerance for the residual.
/// * `initial_guess` - The initial guess for the solution.
///
/// # Tolerance
///
/// The tolerance is the stopping criterion for the CG method. It is the maximum of the absolute tolerance and the relative tolerance.
///
/// The stopping criterion is `|r_k| ≤ max(reltol * |r_0|, abstol)`,
/// where `r_k ≈ A x_k - b` is the residual at the `k`th iteration, and `r_0 ≈ A x_0 - b` is the initial residual.
///
/// # Initial Guess
///
/// The initial guess is the starting point for the CG method. It is used to compute the initial residual.
///
/// # Examples
///
/// ```rust
/// use nalgebra::{DMatrix, DVector};
/// use iterative_solvers::cg_with_initial_guess;
///
/// // Create a simple 2x2 symmetric positive definite system
/// let mat = DMatrix::from_row_slice(2, 2, &[4.0, 1.0, 1.0, 3.0]);
/// let rhs = DVector::from_vec(vec![1.0, 2.0]);
/// let abstol = 1e-10;
/// let reltol = 1e-8;
/// let initial_guess = DVector::from_vec(vec![0.0, 0.0]);
///
/// let solution = cg_with_initial_guess(&mat, &rhs, initial_guess, abstol, reltol).unwrap();
/// ```
pub fn cg_with_initial_guess<'mat>(
    mat: &'mat DMatrix<f64>,
    rhs: &'mat DVector<f64>,
    initial_guess: DVector<f64>,
    abstol: f64,
    reltol: f64,
) -> IterSolverResult<CG<'mat>> {
    let mut solver = CG::new_with_initial_guess(mat, rhs, initial_guess, abstol, reltol)?;
    solver.by_ref().count();
    Ok(solver)
}

#[cfg(test)]
mod tests {
    use std::f64::consts::PI;

    use super::*;
    use crate::utils::symmetric_tridiagonal;

    #[test]
    fn test_cg() {
        let n = 1024;
        let h = 1.0 / 1024.0;
        let a = vec![2.0 / (h * h); n - 1];
        let b = vec![-1.0 / (h * h); n - 2];
        let mat = symmetric_tridiagonal(&a, &b).unwrap();
        let rhs: Vec<_> = (1..n)
            .map(|i| PI * PI * (i as f64 * h * PI).sin())
            .collect();
        let solution: Vec<_> = (1..n).map(|i| (i as f64 * h * PI).sin()).collect();
        let solution = DVector::from_vec(solution);
        let rhs = DVector::from_vec(rhs);
        let solver = cg(&mat, &rhs, 1e-10, 1e-8).unwrap();
        let e = (solution - solver.solution()).norm();
        assert!(e < 1e-4);
    }
}

File: py-iterative-solvers/bench/cg.py
import numpy as np
import scipy.sparse as sp

Mat = np.ndarray[tuple[int, int], np.dtype[np.float64]] | sp.csr_matrix | sp.csc_matrix

Vec = np.ndarray[tuple[int], np.dtype[np.float64]]


class CG:
    """Conjugate Gradient (CG) iterative solver for symmetric positive definite linear systems.

    The Conjugate Gradient method is an iterative algorithm for solving systems of linear
    equations Ax = b where the coefficient matrix A is symmetric and positive definite.
    It is particularly effective for large sparse systems.

    Parameters
    ----------
    mat : ndarray or sparse matrix
        Coefficient matrix A. Must be square, symmetric, and positive definite.
        Supported types: numpy.ndarray, scipy.sparse.csr_matrix, scipy.sparse.csc_matrix.
    rhs : ndarray
        Right-hand side vector b.
    initial_guess : ndarray, optional
        Initial guess for the solution. If None, zero vector is used.
        Default is None.
    abstol : float, optional
        Absolute tolerance for convergence. Default is 1e-10.
    reltol : float, optional
        Relative tolerance for convergence. Default is 1e-8.

    Attributes
    ----------
    mat : Mat
        The coefficient matrix.
    rhs : Vec
        The right-hand side vector.
    solution : Vec
        Current solution estimate.
    residual : float
        Current residual norm.
    iteration : int
        Current iteration number.
    tol : float
        Convergence tolerance (max of abstol and reltol * initial_residual).

    Notes
    -----
    The stopping criterion is ||r_k|| ≤ max(abstol, reltol * ||r_0||),
    where r_k is the residual at iteration k and r_0 is the initial residual.

    The CG method is guaranteed to converge in at most n iterations for an
    n×n symmetric positive definite matrix, though it often converges much
    faster in practice.

    Examples
    --------
    >>> import numpy as np
    >>> import scipy.sparse as sp
    >>> from cg import CG
    >>>
    >>> # Create a symmetric positive definite system
    >>> n = 100
    >>> A = sp.diags([1, -2, 1], [-1, 0, 1], shape=(n, n))
    >>> b = np.random.rand(n)
    >>>
    >>> # Solve using CG
    >>> cg = CG(A, b)
    >>> for iteration, residual, solution in cg:
    ...     if iteration % 10 == 0:
    ...         print(f"Iteration {iteration}, residual: {residual:.2e}")
    """

    def __init__(
        self,
        mat: Mat,
        rhs: Vec,
        initial_guess: Vec | None = None,
        abstol: float = 1e-10,
        reltol: float = 1e-8,
    ):
        if mat.shape[0] != mat.shape[1]:  # type: ignore
            raise ValueError(
                "The matrix is not square, whose shape is ({}, {})",
                mat.shape[0],  # type: ignore
                mat.shape[1],  # type: ignore
            )

        if mat.shape[0] != rhs.shape[0]:  # type: ignore
            raise ValueError(
                "The square matrix with order {}, and the rhs with length {}, do not match",
                mat.shape[0],  # type: ignore
                rhs.shape[0],
            )

        if initial_guess is not None:
            if initial_guess.shape[0] != mat.shape[0]:  # type: ignore
                raise ValueError(
                    "The initial guess with length {}, does not match the matrix with order {}",
                )

        self.mat = mat
        self.rhs = rhs

        if initial_guess is None:
            self.solution = np.zeros_like(rhs)
            self.r = rhs.copy()
        else:
            self.solution = initial_guess
            self.r = rhs - mat.dot(initial_guess)

        self.c = np.zeros_like(rhs)
        self.u = np.zeros_like(rhs)
        self.residual = float(np.linalg.norm(self.r))
        self.prev_residual = self.residual
        self.iteration = 0
        self.tol = max(abstol, reltol * self.residual)

    def converged(self) -> bool:
        """Check if the solver has converged.

        Returns
        -------
        bool
            True if the current residual is below the tolerance threshold.
        """
        return self.residual <= self.tol

    def done(self) -> bool:
        """Check if the iteration should terminate.

        Returns
        -------
        bool
            True if either converged or maximum iterations reached.
        """
        return self.iteration >= self.max_iter() or self.converged()

    def max_iter(self) -> int:
        """Get the maximum number of iterations.

        Returns
        -------
        int
            Maximum number of iterations (equal to problem size).
        """
        return self.solution.shape[0]

    def __iter__(self):
        """Return the iterator object (self).

        Returns
        -------
        CG
            Iterator object for the CG solver.
        """
        return self

    def __next__(self) -> tuple[int, float, Vec]:
        """Perform one iteration of the Conjugate Gradient algorithm.

        Returns
        -------
        tuple[int, float, Vec]
            A tuple containing:
            - iteration : int
                Current iteration number.
            - residual : float
                Current residual norm ||r_k||.
            - solution : Vec
                Current solution estimate x_k.

        Raises
        ------
        StopIteration
            When the algorithm has converged or reached maximum iterations.
        """
        if self.done():
            raise StopIteration

        beta = self.residual**2 / self.prev_residual**2

        self.u *= beta
        self.u += self.r

        self.c[:] = self.mat.dot(self.u)

        alpha = self.residual**2 / np.dot(self.u, self.c)
        self.solution += alpha * self.u
        self.r -= alpha * self.c

        self.prev_residual = self.residual
        self.residual = float(np.linalg.norm(self.r))
        self.iteration += 1

        return self.iteration, self.residual, self.solution


def cg(
    mat: Mat,
    rhs: Vec,
    initial_guess: Vec | None = None,
    abstol: float = 1e-10,
    reltol: float = 1e-8,
) -> tuple[float, Vec]:
    """Solve a linear system using the Conjugate Gradient method.

    This is a convenience function that creates a CG solver and runs it to completion.

    Parameters
    ----------
    mat : ndarray or sparse matrix
        Coefficient matrix A. Must be square, symmetric, and positive definite.
    rhs : ndarray
        Right-hand side vector b.
    initial_guess : ndarray, optional
        Initial guess for the solution. If None, zero vector is used.
    abstol : float, optional
        Absolute tolerance for convergence. Default is 1e-10.
    reltol : float, optional
        Relative tolerance for convergence. Default is 1e-8.

    Returns
    -------
    tuple[float, Vec]
        A tuple containing:
        - residual : float
            Final residual norm.
        - solution : Vec
            Solution vector.
    """
    cg_solver = CG(mat, rhs, initial_guess, abstol, reltol)

    for _ in cg_solver:
        pass

    return cg_solver.residual, cg_solver.solution


if __name__ == "__main__":
    n = 1024
    h = 1.0 / n
    a = [2.0 / (h * h)] * (n - 1)
    b = [-1.0 / (h * h)] * (n - 2)
    mat = np.diag(a) + np.diag(b, k=1) + np.diag(b, k=-1)
    rhs = np.array([np.pi**2 * np.sin(i * h * np.pi) for i in range(1, n)])
    mat = sp.csr_matrix(mat)
    cg_solver = CG(mat, rhs)
    import time

    start = time.time()
    for iteration, residual, _ in cg_solver:
        print(f"Iteration: {iteration}, Residual: {residual}")
    end = time.time()
    print(f"Time: {(end - start) * 1000 * 1000} μs")
